cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,156 @@
+			// so commands can use them in their RunE func
+			newCtx := context.WithValue(cmd.Context(), OtelMeterCtxKey, meter)
+			// newCtx := context.WithValue(newCtx, OtelTracerCtxKey, tracer))
+			cmd.SetContext(newCtx)
+
+			// start the timer as the first step of every cobra command
+			stopCobraCmdTimer := startCobraCommandTimer(cmd)
+			cmdErr := ogRunE(cmd, args)
+			stopCobraCmdTimer(cmdErr)
+			return cmdErr
+		}
+
+		return ogPersistentPreRunE(cmd, args)
+	}
+}
+
+func startCobraCommandTimer(cmd *cobra.Command) func(err error) {
+	ctx := cmd.Context()
+	meter, ok := ctx.Value(OtelMeterCtxKey).(metric.Meter)
+	if meter == nil || !ok {
+		// TODO(krissetto): Handle this better
+		return nil
+	}
+	baseAttrs := BaseMetricAttributes(cmd)
+	durationCounter, _ := meter.Float64Counter(
+		"command.duration",
+		metric.WithDescription("Measures the duration of the cobra command"),
+		metric.WithUnit("ms"),
+	)
+	start := time.Now()
+
+	return func(err error) {
+		duration := float64(time.Since(start)) / float64(time.Millisecond)
+		extraAttrs := []attribute.KeyValue{}
+		exitCode := 0
+		if err != nil {
+			exitCode = 1
+			if stderr, ok := err.(statusError); ok {
+				// StatusError should only be used for errors, and all errors should
+				// have a non-zero exit status, so only set this here if this value isn't 0
+				if stderr.StatusCode != 0 {
+					exitCode = stderr.StatusCode
+				}
+			}
+			// TODO(krissetto): this attribute might not be needed, but buildx has it. Check if it's necessary
+			extraAttrs = append(extraAttrs, attribute.String("command.error.type", otelErrorType(err)))
+		}
+		extraAttrs = append(extraAttrs, attribute.String("command.status", mapExitCode(exitCode)))
+		extraAttrs = append(extraAttrs, attribute.String("command.exit_code", fmt.Sprint(exitCode)))
+		durationCounter.Add(ctx, duration,
+			metric.WithAttributeSet(baseAttrs),
+			metric.WithAttributeSet(attribute.NewSet(extraAttrs...)),
+		)
+	}
+}
+
+func mapExitCode(exitCode int) string {
+	result := "failure"
+
+	if exitCode == 0 {
+		result = "success"
+	}
+
+	return result
+}
+
+// otelErrorType returns an attribute for the error type based on the error category.
+// If nil, this function returns an invalid attribute.
+func otelErrorType(err error) string {
+	name := "generic"
+	if errors.Is(err, context.Canceled) {
+		name = "canceled"
+	}
+	return name
+}
+
+// statusError reports an unsuccessful exit by a command.
+type statusError struct {
+	Status     string
+	StatusCode int
+}
+
+func (e statusError) Error() string {
+	return fmt.Sprintf("Status: %s, Code: %d", e.Status, e.StatusCode)
+}
+
+// getFullCommandName gets the full cobra command name in the format
+// `... parentCommandName commandName` by traversing it's parent commands recursively.
+func getFullCommandName(cmd *cobra.Command) string {
+	cmdName := cmd.Name()
+	if cmd.HasParent() {
+		return fmt.Sprintf("%s %s", getFullCommandName(cmd.Parent()), cmdName)
+	}
+	return cmdName
+}
```

Comment:
Didn't get the time to run this, does this basically do the same thing as Compose is doing here - https://github.com/docker/compose/blob/759d3bc9f9cc401f54753f2384be63aef3b0b2cc/cmd/cmdtrace/cmd_span.go#L125-L135 ?

If so, maybe we can copy-paste it, and the tests (and then maybe later they can reuse this imported from the CLI).

Reply:
It's similar, but it also excludes "docker" (as the root command) and "compose" (as a special case) from the result. I can remove the "docker" (or whatever the root cmd is) part from the result, but i'd honestly leave the "compose" one. 

If we want to share it i could add the possibility of filtering out specific names, but maybe that should just be done case-by-case by whoever wants to exclude specific parts from the output, WDYT?

Reply:
Right, I was thinking that we would exclude `docker` since these are already metrics from the `docker` cli (and we have other attributes to know that).
My logic is that, the same way that their "root command" is `docker compose` and that part is excluded, our root command is `docker` and we could exclude that. But I also don't know how that lines up with parity vs what we had from the wrapper.

Writing some tests for this would also make the expected output of this function easier to find, I'd appreciate it.

Reply:
Not an issue, "docker" will be excluded. 

Regarding compose though, `docker compose` is not their root command and they have specific logic to check if the command name they've reached is == to the plugin name (a const they have). It works for them, but i'm not a huge fan of the approach tbh, as it limits how commands can be named if they aren't the root command (essentially a command cannot have the same name as any of it's parents, or the logic to get the command name will fail). They also have no specific tests for the name retrieval code.

For now I'll push some changes to this code (along with a couple basic unit tests) so that the "first" name (so that of the root command) is excluded, and if you agree we can decide later if it makes sense to also add more logic so that other projects can reuse this (quite basic) code

Reply:
> Regarding compose though, docker compose is not their root command and they have specific logic to check if the command name they've reached is == to the plugin name (a const they have)

Yes, by "root command" (see I had the "s) I meant that that's it's the base command for Docker Compose, as in, the way people use Compose and the command that holds all of their commands. I'd also argue that in general it would be bad to have any command have the same name as it's parent (`docker compose compose`? `docker container container`?) so I wouldn't worry about that.

Reply:
I just took a look at the changes though, looks good :)